src/java/org/apache/cassandra/index/sai/SSTableIndex.java

```diff
@@ -144,7 +147,11 @@ public List<RangeIterator> search(Expression expression,
                                       boolean defer,
                                       int limit) throws IOException
     {
-        return searchableIndex.search(expression, keyRange, context, defer, limit);
+        // for NEQ, NOT_CONTAINS_KEY, NOT_CONTAINS_VALUE we return everything
+        // and AntiJoin + post-filtering at the top level will filter out the unnecesary keys
```

Comment:
I know our goal is to ship this quickly, so this comment is mostly just for future thought and to mention a possible alternative design.

In light of the work I am completing with #804, I think there is a way to push the not operator (whether for equality, contains key, or contains value) down to each segment, at least for inverted indexes. In that scenario, we'd essentially have a list of trie keys to skip and as we iterate over the index, we skip those trie keys while returning the values (segment ids) associated with all other trie keys.

I am not very familiar with the performance profile of that solution compared to this one. Perhaps it is faster to read the Primary Keys sequentially and do the get on the handful of keys we'll skip. Either way, just wanted to mention that I see an alternative path.

Reply:
But how would pushing down work if there are multiple sstables involved? The value that causes the row to be skipped might be in a different sstable than the one we're scanning.

Reply:
> The value that causes the row to be skipped might be in a different sstable than the one we're scanning.

The problem is actually present in both solutions.

In this PR's solution, we get all of the rows that match, map those row ids to PrimaryKeys, union those PrimaryKeys, and then do an anti join against all of the Primary Keys for the table. Problems will arise when an earlier version of the row matches the predicate (therefore removing it from the result set) but a later version of the row does not match because the value was updated. These would be false negatives, results that should have matched but didn't.

In my proposed solution, we would have two issues. You raised the first of incorrectly included Primary Keys. These would be false positives. However, that kind of error is tolerable because we have the row filter to remove rows that matched in the index but don't match when read from storage. The second issue that my solution wouldn't address is the matching of rows with null values for the column.

Here is a test showing the false negative that I described above. This test fails when run on https://github.com/datastax/cassandra/pull/820/commits/674fa05a1aafc90af4ed4dc00243b364cfbc320d 

```java
    @Test
    public void testUpdatedRow() throws Throwable
    {
        createTable("CREATE TABLE %s(id int PRIMARY KEY, text_map map<text, text>)");
        
        execute("INSERT INTO %s(id, text_map) values (1, {'k1':'v1'})");
        
        flush();
        
        execute("INSERT INTO %s(id, text_map) values (1, {'k2':'v2'})");
        
        createIndex("CREATE CUSTOM INDEX ON %s(values(text_map)) USING 'StorageAttachedIndex'");
        
        waitForIndexQueryable();
        
        beforeAndAfterFlush(() -> {
            assertRows(execute("SELECT id FROM %s WHERE text_map NOT CONTAINS 'v1'"), row(1));
        });
    }
```

I am not sure which way to proceed, but I think we must handle this case.

Reply:
One valid solution could be to:

1. Query each sstable for rows that match the negated predicate
2. For each sstable, compute the disjoint union (anti-join) on the results from 1 and on all of the Primary Keys for the table--it is not valid to drop Primary Keys out of the range of the sstable because those PKs might be out of range due to having null values.
3. Union the results of 2.

Given that we do not benefit from PKs out of sstable range, it seems likely we'll have many false positives, a.k.a. shadowed keys.

Reply:
Great catch! Indeed this is a serious fundamental bug. I forgot that we can return false positives from the sstable index due to updates 🤦  And antijoins turn benign false positives into unacceptable false negatives. Ouch. 

So I guess we just need to push the antijoin down and do it only per sstable. We will hae many false positives if a row is split between sstables.

Reply:
> We can convert not equals, not contains, and not contains key into equivalent range queries

I'm not sure how that would work. 
Let's consider a query `SELECT * FROM tab WHERE s NOT CONTAINS 'x'`

Do you mean transforming it into:
`SELECT * FROM tab WHERE s CONTAINS value < 'x' UNION SELECT * FROM tab WHERE s CONTAINS value > 'x' `?

But that's not semantically equivalent, as rows which contain 'x'` may contain other values as well and may match.

Reply:
Great point. I wasn't thinking about collection columns.

Reply:
My only point is to continue the conversation about the design for this feature. The following isn't really meant as something I think we should do, but I spent time researching this, so I want to share my results.

I think that my proposed solution would technically work because lists would produce false positives and those are acceptable from SAI iterators. Here is a test that works correctly on your branch and shows that the contains problem you describe is present even in your current solution:

```java
    @Test
    public void notContainsForSplitUnfrozenList() throws Throwable
    {
        createTable("CREATE TABLE %s(id int PRIMARY KEY, texts list<text>)");
        createIndex("CREATE CUSTOM INDEX ON %s(texts) USING 'StorageAttachedIndex'");
        waitForIndexQueryable();
        execute("INSERT INTO %s(id, texts) values (1, ?)", list("v1"));
        flush();

        // This adds to, but doesn't overwrite, v1
        execute("UPDATE %s SET texts = texts + [ 'v2' ] WHERE id=1");

        beforeAndAfterFlush(() -> {
            assertRows(execute("SELECT id FROM %s WHERE texts NOT CONTAINS 'v1'"));
        });
    }
```

If you look with a debugger at the execution of the test, the SAI `RangeIterator` produces row 1, but it is filtered out when we verify that the row's representation from storage is a valid match.

I think we would see the same behavior if we mapped NOT CONTAINS to a range. However, that would come with additional, like implementing null indexing, and it might also lead to more false positives. The primary theoretical benefit is skipping the "read all PKs for table" step. That might not actually be better than doing a range through the trie though, considering we'd read all of the PKs in order out of the file.